{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"https://github.com/Colorado-State-University-CMB/CM515-course-2024/blob/main/modules/14_Google_Colab/Module14_Homework.ipynb","timestamp":1714865589081}]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":19,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"jviYcubNtZdz","outputId":"0f3c61d5-668d-404a-f5f0-28afc9ad9b52","executionInfo":{"status":"ok","timestamp":1714910138418,"user_tz":360,"elapsed":120,"user":{"displayName":"Nneka Okereke","userId":"12602915177121695742"}}},"outputs":[{"output_type":"execute_result","data":{"text/plain":["[2, 4, 6, 4, 10, 12, 6, 4, 2]"]},"metadata":{},"execution_count":19}],"source":["## Homework Questions\n","# 1. Convert the following code into a loop\n","\n","x = [1,1,2,3,2,5,6,7,8]\n","\n","x[8] = x[0] * 2\n","x[0] = x[1] * 2\n","x[1] = x[2] * 2\n","x[2] = x[3] * 2\n","x[3] = x[4] * 2\n","x[4] = x[5] * 2\n","x[5] = x[6] * 2\n","x[7] = x[8] * 2\n","\n","x"]},{"cell_type":"code","source":["x = [1, 1, 2, 3, 2, 5, 6, 7, 8]\n","\n","# Loop through the indices 0 to 7\n","for i in range(8):\n","    x[i] = x[i + 1] * 2\n","\n","# Handling the special case for the last element\n","x[8] = x[0] * 2\n","\n","print(x)\n","\n","#cant seem to get the exact output :(\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"rOlUG17nmqb5","executionInfo":{"status":"ok","timestamp":1714911548824,"user_tz":360,"elapsed":137,"user":{"displayName":"Nneka Okereke","userId":"12602915177121695742"}},"outputId":"4ef0ef94-afe8-41c4-b953-85d99cb053cf"},"execution_count":23,"outputs":[{"output_type":"stream","name":"stdout","text":["[2, 4, 6, 4, 10, 12, 14, 16, 4]\n"]}]},{"cell_type":"code","source":["# 2. load in the GAPDH template strant from the fasta provided here https://github.com/Colorado-State-University-CMB/CM515-course-2024/tree/main/modules/14_Google_Colab\n","#!git clone https://github.com/Colorado-State-University-CMB/CM515-course-2024.git\n","path = \"/content/CM515-course-2024/modules/14_Google_Colab/\"\n","\n","with open(path + 'gapdh.fasta', 'r') as file:\n","  file_list = file.readlines()\n","file_list\n"],"metadata":{"id":"UjGLWNLqt29c","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1714908306320,"user_tz":360,"elapsed":151,"user":{"displayName":"Nneka Okereke","userId":"12602915177121695742"}},"outputId":"d7db2834-af86-4386-818e-408da182ae8e"},"execution_count":9,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['>mm_GAPDH_template_strand\\n',\n"," 'GGTGGGTGCAGCGAACTTTATTGATGGTATTCAAGAGAGTAGGGAGGGCTCCCTAGGCCCCTCCTGTTATTATGGGGGTCTGGGATGGAAATTGTGAGGGAGATGCTCAGTGTTGGGGGCCGAGTTGGGATAGGGCCTCTCTTGCTCAGTGTCCTTGCTGGGGTGGGTGGTCCAGGGTTTCTTACTCCTTGGAGGCCATGTAGGCCATGAGGTCCACCACCCTGTTGCTGTAGCCGTATTCATTGTCATACCAGGAAATGAGCTTGACAAAGTTGTCATTGAGAGCAATGCCAGCCCCGGCATCGAAGGTGGAAGAGTGGGAGTTGCTGTTGAAGTCGCAGGAGACAACCTGGTCCTCAGTGTAGCCCAAGATGCCCTTCAGTGGGCCCTCAGATGCCTGCTTCCCATTCTCGGCCTTGACTGTGCCGTTGAATTTGCCGTGAGTGGAGTCATACTGGAACATGTAGACCATGTAGTTGAGGTCAATGAAGGGGTCGTTGATGGCAACAATCTCCACTTTGCCACTGCAAATGGCAGCCCTGGTGACCAGGCGCCCAATACGGCCAAATCCGTTCACACCGACCTTCACCATTTTGTCCCTCGTGCC']"]},"metadata":{},"execution_count":9}]},{"cell_type":"code","source":["# 3. calculate the GC content of the strant (report as a percenatage)\n","def calculate_gc_content(seq):\n","    gc_count = seq.count('G') + seq.count('C')\n","    total_bases = len(seq)\n","    gc_content = (gc_count / total_bases) * 100\n","    return gc_content\n","\n","gc_content = calculate_gc_content(file_list[1])\n","print(gc_content)"],"metadata":{"id":"rDTwXprCt3BC","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1714908610958,"user_tz":360,"elapsed":162,"user":{"displayName":"Nneka Okereke","userId":"12602915177121695742"}},"outputId":"2b3f8851-4570-4d7d-fc3e-dcbd056acc3e"},"execution_count":11,"outputs":[{"output_type":"stream","name":"stdout","text":["54.36573311367381\n"]}]},{"cell_type":"code","source":["# 4. Create a function that takes in a dna sequence and finds the reverse compliment of it\n","def reverse_complement(dna_seq):\n","    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n","    reverse_comp_seq = ''.join([complement[base] for base in reversed(dna_seq)])\n","    return reverse_comp_seq\n","\n","print(reverse_complement(file_list[1]))"],"metadata":{"id":"j-3QxgN9t3Es","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1714908660614,"user_tz":360,"elapsed":128,"user":{"displayName":"Nneka Okereke","userId":"12602915177121695742"}},"outputId":"9bb758ce-bced-48db-9d5c-88a40a06e14e"},"execution_count":12,"outputs":[{"output_type":"stream","name":"stdout","text":["GGCACGAGGGACAAAATGGTGAAGGTCGGTGTGAACGGATTTGGCCGTATTGGGCGCCTGGTCACCAGGGCTGCCATTTGCAGTGGCAAAGTGGAGATTGTTGCCATCAACGACCCCTTCATTGACCTCAACTACATGGTCTACATGTTCCAGTATGACTCCACTCACGGCAAATTCAACGGCACAGTCAAGGCCGAGAATGGGAAGCAGGCATCTGAGGGCCCACTGAAGGGCATCTTGGGCTACACTGAGGACCAGGTTGTCTCCTGCGACTTCAACAGCAACTCCCACTCTTCCACCTTCGATGCCGGGGCTGGCATTGCTCTCAATGACAACTTTGTCAAGCTCATTTCCTGGTATGACAATGAATACGGCTACAGCAACAGGGTGGTGGACCTCATGGCCTACATGGCCTCCAAGGAGTAAGAAACCCTGGACCACCCACCCCAGCAAGGACACTGAGCAAGAGAGGCCCTATCCCAACTCGGCCCCCAACACTGAGCATCTCCCTCACAATTTCCATCCCAGACCCCCATAATAACAGGAGGGGCCTAGGGAGCCCTCCCTACTCTCTTGAATACCATCAATAAAGTTCGCTGCACCCACC\n"]}]},{"cell_type":"code","source":["# 5. Convert the GAPDH template strand to the coding strand\n","def coding_strand(dna_sequence):\n","    rna_sequence = dna_sequence.replace('T', 'U')\n","    coding_sequence = rna_sequence.replace('A', 'T').replace('C', 'G').replace('G', 'C')\n","    return coding_sequence\n","\n","print(coding_strand(file_list[1]))"],"metadata":{"id":"Zmgg9BzGt3IK","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1714908712583,"user_tz":360,"elapsed":166,"user":{"displayName":"Nneka Okereke","userId":"12602915177121695742"}},"outputId":"20740427-fc0f-4623-b7d8-f8ef33d63b23"},"execution_count":13,"outputs":[{"output_type":"stream","name":"stdout","text":["CCUCCCUCCTCCCTTCUUUTUUCTUCCUTUUCTTCTCTCUTCCCTCCCCUCCCUTCCCCCCUCCUCUUTUUTUCCCCCUCUCCCTUCCTTTUUCUCTCCCTCTUCCUCTCUCUUCCCCCCCCTCUUCCCTUTCCCCCUCUCUUCCUCTCUCUCCUUCCUCCCCUCCCUCCUCCTCCCUUUCUUTCUCCUUCCTCCCCTUCUTCCCCTUCTCCUCCTCCTCCCUCUUCCUCUTCCCCUTUUCTUUCUCTUTCCTCCTTTUCTCCUUCTCTTTCUUCUCTUUCTCTCCTTUCCCTCCCCCCCCTUCCTTCCUCCTTCTCUCCCTCUUCCUCUUCTTCUCCCTCCTCTCTTCCUCCUCCUCTCUCUTCCCCTTCTUCCCCUUCTCUCCCCCCUCTCTUCCCUCCUUCCCTUUCUCCCCCUUCTCUCUCCCCUUCTTUUUCCCCUCTCUCCTCUCTUTCUCCTTCTUCUTCTCCTUCUTCUUCTCCUCTTUCTTCCCCUCCUUCTUCCCTTCTTUCUCCTCUUUCCCTCUCCTTTUCCCTCCCCUCCUCTCCTCCCCCCCTTUTCCCCCTTTUCCCUUCTCTCCCTCCUUCTCCTUUUUCUCCCUCCUCCC\n"]}]},{"cell_type":"code","source":["# 6. convert to the mRNA sequence\n","def dna_to_mRNA(dna_sequence):\n","    return dna_sequence.replace('T', 'U')\n","\n","print(dna_to_mRNA(file_list[1]))\n","\n","mRNA = dna_to_mRNA(file_list[1])"],"metadata":{"id":"kOCFY42Qt3K8","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1714909833233,"user_tz":360,"elapsed":147,"user":{"displayName":"Nneka Okereke","userId":"12602915177121695742"}},"outputId":"3b8314fe-c875-4d3a-c0fc-6bd6ff671f8a"},"execution_count":16,"outputs":[{"output_type":"stream","name":"stdout","text":["GGUGGGUGCAGCGAACUUUAUUGAUGGUAUUCAAGAGAGUAGGGAGGGCUCCCUAGGCCCCUCCUGUUAUUAUGGGGGUCUGGGAUGGAAAUUGUGAGGGAGAUGCUCAGUGUUGGGGGCCGAGUUGGGAUAGGGCCUCUCUUGCUCAGUGUCCUUGCUGGGGUGGGUGGUCCAGGGUUUCUUACUCCUUGGAGGCCAUGUAGGCCAUGAGGUCCACCACCCUGUUGCUGUAGCCGUAUUCAUUGUCAUACCAGGAAAUGAGCUUGACAAAGUUGUCAUUGAGAGCAAUGCCAGCCCCGGCAUCGAAGGUGGAAGAGUGGGAGUUGCUGUUGAAGUCGCAGGAGACAACCUGGUCCUCAGUGUAGCCCAAGAUGCCCUUCAGUGGGCCCUCAGAUGCCUGCUUCCCAUUCUCGGCCUUGACUGUGCCGUUGAAUUUGCCGUGAGUGGAGUCAUACUGGAACAUGUAGACCAUGUAGUUGAGGUCAAUGAAGGGGUCGUUGAUGGCAACAAUCUCCACUUUGCCACUGCAAAUGGCAGCCCUGGUGACCAGGCGCCCAAUACGGCCAAAUCCGUUCACACCGACCUUCACCAUUUUGUCCCUCGUGCC\n"]}]},{"cell_type":"code","source":["# 7. Translate the mRNA\n","# make a function that finds the start codon (AUG), then translates the protein until a stop codon is reached.\n","# stop codons are denoted as \"*\" in the codon table file provided\n","\n","def translate_dna(sequence):\n","  #Create a dictionary representing codons and amino acids https://en.wikipedia.org/wiki/DNA_and_RNA_codon_tables\n","    genetic_code = {\n","        'AUA':'I', 'AUC':'I', 'AUU':'I', 'AUG':'M',\n","        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACU':'T',\n","        'AAC':'N', 'AAU':'N', 'AAA':'K', 'AAG':'K',\n","        'AGC':'S', 'AGU':'S', 'AGA':'R', 'AGG':'R',\n","        'CUA':'L', 'CUC':'L', 'CUG':'L', 'CUU':'L',\n","        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCU':'P',\n","        'CAC':'H', 'CAU':'H', 'CAA':'Q', 'CAG':'Q',\n","        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGU':'R',\n","        'GUA':'V', 'GUC':'V', 'GUG':'V', 'GUU':'V',\n","        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCU':'A',\n","        'GAC':'D', 'GAU':'D', 'GAA':'E', 'GAG':'E',\n","        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGU':'G',\n","        'UCA':'S', 'UCC':'S', 'UCG':'S', 'UCU':'S',\n","        'UUC':'F', 'UUU':'F', 'UUA':'L', 'UUG':'L',\n","        'UAC':'Y', 'UAU':'Y', 'UAA':'*', 'UAG':'*',\n","        'UGC':'C', 'UGU':'C', 'UGA':'*', 'UGG':'W',\n","    }\n","\n","    start_codon = 'AUG'\n","    stop_codons = ['UAA', 'UAG', 'UGA']\n","\n","    protein = \"\"\n","    for i in range(0, len(sequence), 3):\n","        codon = sequence[i:i+3]\n","        if codon == start_codon:\n","            for j in range(i, len(sequence), 3):\n","                codon = sequence[j:j+3]\n","                if codon in stop_codons:\n","                    return protein\n","                amino_acid = genetic_code.get(codon, '*')\n","                protein += amino_acid\n","    return protein\n","print(translate_dna(mRNA))\n"],"metadata":{"id":"cUEL3pVqt3PJ","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1714909981022,"user_tz":360,"elapsed":121,"user":{"displayName":"Nneka Okereke","userId":"12602915177121695742"}},"outputId":"33ece4a2-0864-4377-fbd3-d2a488342ce3"},"execution_count":17,"outputs":[{"output_type":"stream","name":"stdout","text":["MEIVREMLSVGGRVGIGPLLLSVLAGVGGPGFLTPWRPCRP\n"]}]},{"cell_type":"code","source":["# 8. Go to this website, select 2 problems and complete them https://rosalind.info/problems/list-view/"],"metadata":{"id":"K6jGkxf2t3Sh"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Counting nucleotides in a seqence (Question 1)\n","sequence = \"TGTCAAAGCCACTATCTTAGGTGAGATTGAATTCGGCAGGATGATCTAGGGCGGAGGCCAGCATTGCGAAATTGCTAGTACTGATTGACTGGAGGAGGCAAGTGGGAATACTTAGCAAGGTCATCGGTTGGATGCCACGGAGCCGCTTATATGCATATTAGTACCCGAGAGTCTAGCACAGCTATCACGGGCCTGGGCGTATCCATATAATCCACTCGGGAGCTCGTCACCCGCCCGCGATACGCAAGGATGCGGACGGCCCGGAAGGCAAGCGTGATCCCAGACTTATTAGCCACTTGTCGTACCCAACATGCGTCCATTGATGGGGAGCGATCAGCAAGGTGACTAGCAAGATCTGCCAGGACTTTCTGCTCTTTATCCGCTCGTTTTCTTCTTGTAATCCAGAGTACAGCCGTATTATGCGCTTCCTTTTATGTATGCTCATATCGCAGGGTACCACCTTTCGAATCGCTATCCACATGTGGCGTCCAAATCAGTCCGCCGTGGGCCGACGTTGATTCCTCAAATATTGGGACGGTTTGCGACAAGAGGGGCCAGTACTTAACACGTATACGTCCAACCCACCCGCGTGCATTCTAGTCAACTACTGTACTAGGTTTATTGCCCGACGATTCTACGGATCTCCGGTTAGGTTATAATAAGGGTGGCAGAATCTTGTACTGTGTGGTCTACGCGCGACTCACAAAACTTTCGACTTGTTAGCCTCTTAAAAGGTAATATGACACGCTGCTAGAGCTATTAACTCTTAGGCCCAGAAAACCCAAGAGACACACGATTATCAATGAGAGATTCAACTGGAGCGATTCTCTTATGGATTATGAGCGACGTGTAGAACTAATACTCTTGTGCGAGATAATTGAAGCTGGATACCGGGTACCGCTAATACTCATGGATCCACCTTTGAAATGAATGACATGAACACCGTCAGTGTATGCACCCTAGC\"\n","\n","# Initialize counts for each nucleotide\n","counts = {'A': 0, 'C': 0, 'G': 0, 'T': 0}\n","\n","# Count occurrences of each nucleotide\n","for nucleotide in sequence:\n","    counts[nucleotide] += 1\n","\n","# Output the counts as four integers separated by spaces\n","print(counts['A'], counts['C'], counts['G'], counts['T'])\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"bn0uHYd-q-Fv","executionInfo":{"status":"ok","timestamp":1714911187069,"user_tz":360,"elapsed":171,"user":{"displayName":"Nneka Okereke","userId":"12602915177121695742"}},"outputId":"1510e73a-f5dc-454f-f1b6-3b3c219d87f5"},"execution_count":21,"outputs":[{"output_type":"stream","name":"stdout","text":["247 234 237 246\n"]}]},{"cell_type":"code","source":["#Counting Single nucleotide polymorphisms or point mutations (Question 6)\n","\n","seq1 = \"GAGCCTACTAACGGGAT\"\n","seq2 = \"CATCGTAATGACGGCCT\"\n","\n","# Initialize mutation count\n","mutations = 0\n","\n","# Iterate over corresponding positions in the sequences\n","for i in range(len(seq1)):\n","    # Compare nucleotides at each position\n","    if seq1[i] != seq2[i]:\n","        mutations += 1\n","\n","# Output the number of point mutations\n","print(\"Number of point mutations:\", mutations)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"8IF-IAFArrj3","executionInfo":{"status":"ok","timestamp":1714911402439,"user_tz":360,"elapsed":130,"user":{"displayName":"Nneka Okereke","userId":"12602915177121695742"}},"outputId":"4dd91f62-9438-44e1-c151-ccb98068747b"},"execution_count":22,"outputs":[{"output_type":"stream","name":"stdout","text":["Number of point mutations: 7\n"]}]}]}